### PR TITLE
Concurrent Transaction Conditonal Creates

### PR DESCRIPTION
**Pull request by AD1306**
Avoid a concurrency failure when executing transactions at high concurrency using conditional creates and placeholder references

### CODE REVIEW

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ConcurrentWriteTest.java [hunk 1]
```diff
@@ -162,9 +191,9 @@ public void testTransactionCreates_WithRetry() throws ExecutionException, Interr
 
 			assertEquals(1, counts.get("Patient"), counts.toString());
 			assertEquals(1, counts.get("Observation"), counts.toString());
-			assertEquals(6, myResourceLinkDao.count());
-			assertEquals(6, myResourceTableDao.count());
-			assertEquals(14, myResourceHistoryTableDao.count());
+			assertEquals(7, myResourceLinkDao.count()); // 1 for SP, 6 for transaction
+			assertEquals(8, myResourceTableDao.count()); // 2 SPs, 6 resources
```
Comment (TipzCM) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ConcurrentWriteTest.java:195]: love the comments

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java [hunk 2]
```diff
@@ -4076,6 +4082,230 @@ public void testTransactionWithPlaceholderIdInMatchUrlPut() {
...
+		});
+
+
+		// execute
+		Bundle output = mySystemDao.transaction(mySrd, bundle);
+
+		// validate
+
+		String observationId = new IdType(output.getEntry().get(0).getResponse().getLocation()).toUnqualifiedVersionless().getValue();
+		String patientId = new IdType(output.getEntry().get(1).getResponse().getLocation()).toUnqualifiedVersionless().getValue();
```
Comment (TipzCM) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java:4136]: Nitpick: should probably start with a validation that `output.getEntry() >= 2`. That way if this does ever fail, it's not an index out of bounds error, but the actual test erroring out.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java [hunk 3]
```diff
@@ -4076,6 +4082,230 @@ public void testTransactionWithPlaceholderIdInMatchUrlPut() {
...
+		ourLog.info("Assigned resource IDs:\n * " + output.getEntry().stream().map(t->t.getResponse().getLocation()).collect(Collectors.joining("\n * ")));
+
+		myCaptureQueriesListener.logInsertQueries(t -> t.getSql(false, false).contains(" into HFJ_RESOURCE "));
+
+		runInTransaction(()->{
+			List<ResourceTable> all = myResourceTableDao.findAll();
+			List<String> storedTypes = all.stream().map(ResourceTable::getResourceType).sorted().toList();
+			assertThat("Resources:\n * " + all.stream().map(t->t.toString()).collect(Collectors.joining("\n * ")), storedTypes, contains("Observation", "Patient"));
+		});
+	}
```
Comment (TipzCM) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java:4306]: These tests are all so similar. I feel like they can be combined into 1 test.

The first 2, for sure, only differ only in the observations subject and the final verify.

These can be a ParameterizedTest with the ValueSource being the subject identifier urls ("http://acme.org", "http://nothing-matching", "") - these URLs can be escaped using `UrlUtils.escapeParam` when setting them to the actual subject line.

Then the last verify line (checking if 1 or 2 patients are created) can be done by checking if the subject does indeed match the patient's system URL to determine which validation to use.

**Review by TipzCM — APPROVED** [review]

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java:4136]: done

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java:4171]: This is a great suggestion, changed as suggested

**Pull request merged**